### PR TITLE
WeightFree export review comment addressed

### PR DESCRIPTION
**Pull request by amarquic**
Summary

  This PR extends the weight-free export pipeline for CausalLM models with CCL support, fixes several export-path bugs, and
  rounds out the weight-free test suite.

  Changes

  Export pipeline

  - utils/cache.py: Add QEFF_CHECKPOINT_HOME env var — controls where prepared checkpoints (dtype-converted, MoE-restacked)
  - utils/torch_patches.py: Add dynamo_invoke_subgraph_fallback_env() context manager that sets
  TORCH_INVOKE_ALLOW_CREATE_FALLBACK=1 for dynamo export; simplify setattr(module, "forward", ...) to direct assignment
  - utils/export_utils.py: Fix dynamo flag inference in export_wrapper (removed incorrect use_weight_free_export coercion);
  simplify nested with context managers
  - base/modeling_qeff.py, exporter/onnx_exporter.py, exporter/weight_free/export.py,
  exporter/weight_free/checkpoint_key_resolver.py, transformers/models/modeling_auto.py: Weight-free export pipeline updates
  including tied-weight alias handling and CCL integration
  
  Test suite
  - tests/weight_free/test_ccl.py (new): Weight-free + CCL compile/generate tests — exports once, compiles twice (auto-CCL and
   explicit comp_ctx_lengths_prefill/decode)
  - tests/weight_free/conftest.py: Mark gpt_oss as xfail across all weight-free tests (SerdeError with dynamo=True + 
  use_onnx_subfunctions=True); restructure pytest_collection_modifyitems for single-pass marker application
  - tests/weight_free/_helpers.py, test_export.py, test_continuous_batching.py, test_on_qaic.py: Extended coverage and helper
  updates

### CODE REVIEW

File: QEfficient/transformers/models/modeling_auto.py [hunk 1]
```diff
@@ -3797,15 +3838,14 @@ def _factory(model_id, config):
 
     def export(
         self,
-        export_dir: Optional[str] = None,
```
Comment (quic-amitraj) [inline QEfficient/transformers/models/modeling_auto.py:3805]: Why these changes?

File: QEfficient/base/modeling_qeff.py [hunk 2]
```diff
@@ -178,7 +177,7 @@ def __init__(self, model: torch.nn.Module, **kwargs) -> None:
 
         # Flag for checking if weights are offloaded
         self._is_weights_offloaded: bool = False
-
+        self._weight_free: bool = False
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: If we already have this, why are we doing `instance._weight_free=True` in `modeling_auto.py` We should pass it via `init`, and let `init` set its value correctly

File: QEfficient/utils/export_utils.py
Comment (ochougul) [inline QEfficient/utils/export_utils.py:1]: update `use_weight_free_export` to  use `qeff_model._weight_free` here.

File: QEfficient/base/modeling_qeff.py [hunk 3]
```diff
@@ -424,12 +422,12 @@ def _export(
             self.weight_spec_path = str(_weight_spec_path) if _weight_spec_path.is_file() else None
             return onnx_path
 
-        if use_weight_free_export:
+        if self._weight_free:
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: you are already doing it in `modeling_auto` why is this logic written at two places? better always do it in `_export`

File: QEfficient/base/modeling_qeff.py [hunk 4]
```diff
@@ -504,7 +502,7 @@ def _resolve_pkv_names(layer_idx, layer_state):
             input_names = aligned_input_names
 
         try:
-            if use_weight_free_export:
+            if self._weight_free:
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py:499]: This is cleaner 
```
            if self._weight_free:
                from QEfficient.exporter.onnx_exporter import export_via_weightfree

                export_result = export_via_weightfree(
                    self,
                    onnx_path,
                    example_inputs,
                    input_names,
                    output_names,
                    dynamic_shapes,
                    export_kwargs,
                    onnx_transform_kwargs,
                )
            elif dynamo:
                from QEfficient.exporter.onnx_exporter import export_via_dynamo
                self._model_offloaded_check()
                export_result = export_via_dynamo(
                    self,
                    onnx_path,
                    example_inputs,
                    input_names,
                    output_names,
                    dynamic_shapes,
                    export_kwargs,
                )
                self._offload_model_weights(offload_pt_weights)
            else:
                from QEfficient.exporter.onnx_exporter import export_via_legacy
                self._model_offloaded_check()
                export_result = export_via_legacy(
                    self,
                    onnx_path,
                    example_inputs,
                    input_names,
                    output_names,
                    dynamic_axes,
                    export_kwargs,
                )
                self._offload_model_weights(offload_pt_weights)
```

File: QEfficient/base/modeling_qeff.py [hunk 5]
```diff
@@ -424,12 +422,12 @@ def _export(
...
             return onnx_path
 
-        if use_weight_free_export:
+        if self._weight_free:
             dynamo = True
 
         # check if the model is in meta state or weights are offloaded
         # (skip for weight-free export which intentionally uses a meta-device model)
-        if not use_weight_free_export:
+        if not self._weight_free:
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: remove as per below code

File: QEfficient/base/modeling_qeff.py [hunk 6]
```diff
@@ -844,6 +840,7 @@ def _resolve_pkv_names(layer_idx, layer_state):
                     input_names.append(param)
         dynamic_axes = {k: v for k, v in dynamic_axes.items() if k in input_names}
 
+        import os
         import time
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: move to top

File: QEfficient/exporter/onnx_exporter.py [hunk 7]
```diff
@@ -136,59 +129,54 @@ def export_via_weightfree(
     tmp_onnx_path = tmp_onnx_dir / onnx_path.name
     tmp_onnx_dir.mkdir(parents=True, exist_ok=True)
```
Comment (ochougul) [inline QEfficient/exporter/onnx_exporter.py]: this tmp dir is not needed. remove it.

File: QEfficient/exporter/onnx_exporter.py [hunk 8]
```diff
@@ -136,59 +129,54 @@ def export_via_weightfree(
...
+                onnx_transform_kwargs=updated_onnx_transform_kwargs or {},
+                excluded_onnx_transforms=(SplitTensorsTransform,),
+                weight_spec_path=weight_spec_path,
+                post_transform_hooks=post_transform_hooks,
+                post_export_hooks=post_export_hooks,
+            )
+        finally:
+            if cleanup is not None:
+                cleanup()
+            shutil.rmtree(tmp_onnx_dir, ignore_errors=True)
```
Comment (ochougul) [inline QEfficient/exporter/onnx_exporter.py]: remove `try/finally` block, and we can move the `finally` dir removal to the `modeling_qeff` level `try/except`.
Also, this `tmp_dir` is not needed

File: QEfficient/exporter/onnx_exporter.py [hunk 9]
```diff
@@ -17,37 +17,37 @@
...
-    onnx_transform_kwargs: Dict[str, Any] = field(default_factory=dict)
-    excluded_onnx_transforms: Tuple[Type[Any], ...] = ()
-    weight_spec_path: Optional[Path] = None
-    post_transform_hooks: Tuple[Callable[[Any], None], ...] = ()
-    post_export_hooks: Tuple[Callable[[], None], ...] = ()
+    onnx_transform_kwargs: dict[str, Any] = field(default_factory=dict)
+    excluded_onnx_transforms: tuple[type[Any], ...] = ()
+    weight_spec_path: Path | None = None
+    post_transform_hooks: tuple[Callable[[Any], None], ...] = ()
+    post_export_hooks: tuple[Callable[[], None], ...] = ()
```
Comment (ochougul) [inline QEfficient/exporter/onnx_exporter.py]: why is Typehint changed? revert

File: QEfficient/exporter/onnx_exporter.py [hunk 10]
```diff
@@ -136,59 +129,54 @@ def export_via_weightfree(
     tmp_onnx_path = tmp_onnx_dir / onnx_path.name
     tmp_onnx_dir.mkdir(parents=True, exist_ok=True)
 
-    prev_invoke_fallback = os.environ.get("TORCH_INVOKE_ALLOW_CREATE_FALLBACK")
-    os.environ["TORCH_INVOKE_ALLOW_CREATE_FALLBACK"] = "1"
     cleanup = None
```
Comment (ochougul) [inline QEfficient/exporter/onnx_exporter.py]: remove this? its not doing anything

File: QEfficient/utils/torch_patches.py [hunk 11]
```diff
@@ -417,3 +418,23 @@ def temporarily_disable_nested_compile_regions(model, target_classes=None):
                 delattr(module, "forward")
             else:
                 setattr(module, "forward", previous_forward)
+
+
+@contextmanager
```
Comment (ochougul) [inline QEfficient/utils/torch_patches.py:427]: make it thread-safe
```

  _DYNAMO_ENV_LOCK = threading.RLock()

  @contextmanager
  def dynamo_invoke_subgraph_fallback_env():
      with _DYNAMO_ENV_LOCK:
          previous = os.environ.get("TORCH_INVOKE_ALLOW_CREATE_FALLBACK")
          os.environ["TORCH_INVOKE_ALLOW_CREATE_FALLBACK"] = "1"
          try:
              yield
          finally:
              if previous is None:
                  os.environ.pop("TORCH_INVOKE_ALLOW_CREATE_FALLBACK", None)
              else:
                  os.environ["TORCH_INVOKE_ALLOW_CREATE_FALLBACK"] = previous
```

File: QEfficient/utils/cache.py [hunk 12]
```diff
@@ -15,3 +15,7 @@
     QEFF_HOME = Path(os.environ["XDG_CACHE_HOME"]) / "qeff_models"
 else:
     QEFF_HOME = Path("~/.cache/qeff_models").expanduser()
+
+QEFF_CHECKPOINT_HOME: Path = None
+if "QEFF_CHECKPOINT_HOME" in os.environ:
+    QEFF_CHECKPOINT_HOME = Path(os.environ["QEFF_CHECKPOINT_HOME"])
```
Comment (ochougul) [inline QEfficient/utils/cache.py]: `QEFF_CHECKPOINT_HOME = Path(os.environ.get("QEFF_WF_HOME"), None)`

File: QEfficient/exporter/weight_free/export.py [hunk 13]
```diff
@@ -62,8 +37,24 @@ def _to_meta(value: Any) -> Any:
...
+    load time (modeling_auto.py imports from this package's sibling module).
+    """
+    from QEfficient.transformers.models.modeling_auto import _resolve_torch_dtype
+
+    kwargs = {"torch_dtype": model_dtype or torch.float32}
+    _resolve_torch_dtype(kwargs)
+    return kwargs["torch_dtype"]
+
+
 def _build_meta_qeff_model(qeff_model):
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/export.py]: this method should be named `run_quantizer_for_wf`

File: QEfficient/exporter/weight_free/export.py [hunk 14]
```diff
@@ -5,50 +5,25 @@
...
-    temporarily_disable_nested_compile_regions,
-    temporarily_enable_nested_compile_regions,
-)
 
-_PREPARED_CHECKPOINT_LAYOUT_VERSION = "moe-weights-v1"
-
-_last_prep_peak_rss_mb: Optional[float] = None
+_last_prep_peak_rss_mb: float | None = None
 _last_prep_duration_seconds: float = 0.0
 _checkpoint_prep_ran: bool = False
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/export.py]: remove these global vars, not needed

Comment (ochougul) [reply QEfficient/exporter/weight_free/export.py]: Remove all three globals and use a local duration:
```
  prep_start = time.perf_counter()
  prepared_model_ref = _prepare_checkpoint_for_weight_free_export(
      meta_qeff_model,
      model_ref,
      target_dtype,
  )
  prep_duration_seconds = time.perf_counter() - prep_start

  logger.info(
      "Weight-free checkpoint preparation completed in %.2fs: %s",
      prep_duration_seconds,
      prepared_model_ref,
  )
```

File: QEfficient/exporter/weight_free/export.py [hunk 15]
```diff
@@ -129,10 +99,13 @@ def _build_meta_qeff_model(qeff_model):
...
         with init_empty_weights():
-            quantizer._process_model_before_weight_loading(meta_qeff_model.model)
+            quantizer._process_model_before_weight_loading(qeff_model.model)
+    else:
+        target_dtype = _resolve_weight_free_target_dtype(getattr(qeff_model.model.config, "dtype", None))
+        qeff_model.model = qeff_model.model.to(dtype=target_dtype)
 
-    meta_qeff_model.model.eval()
-    return meta_qeff_model
+    qeff_model.model.eval()
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/export.py]: remove this `eval` we are already doing `requires_grad=False` ahead

File: QEfficient/exporter/weight_free/export.py [hunk 16]
```diff
@@ -246,73 +222,62 @@ def export_weight_free_onnx(
...
-            kwargs=meta_example_inputs,
-            input_names=input_names,
-            output_names=output_names,
-            dynamic_axes=None,
-            dynamic_shapes=dynamic_shapes,
-            **export_kwargs,
-        )
-        if onnx_program is None:
-            raise RuntimeError("torch.onnx.export returned None for weight-free dynamo export")
+    onnx_program = torch.onnx.export(
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/export.py]: use the contextmanager `dynamo_invoke_subgraph_fallback_env` for the `torch.onnx.export` call. remove it from the top level call

File: QEfficient/exporter/weight_free/export.py [hunk 17]
```diff
@@ -246,73 +222,62 @@ def export_weight_free_onnx(
...
+        output_names=output_names,
+        dynamic_axes=None,
+        dynamic_shapes=dynamic_shapes,
+        **export_kwargs,
+    )
+    if onnx_program is None:
+        raise RuntimeError("torch.onnx.export returned None for weight-free dynamo export")
 
-        target_dtype = _resolve_weight_free_dtype(qeff_model.model.config)
+    target_dtype = _resolve_weight_free_target_dtype(getattr(qeff_model.model.config, "dtype", None))
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/export.py]: `_resolve_weight_free_target_dtype()` should be removed. The `dtype` was already normalized by
  `_resolve_torch_dtype(kwargs)` in `from_pretrained()`, and `_build_meta_model()` writes that value
  into both `config.dtype` and `config.torch_dtype`.

File: QEfficient/exporter/weight_free/export.py [hunk 18]
```diff
@@ -246,73 +222,62 @@ def export_weight_free_onnx(
     global _checkpoint_prep_ran, _last_prep_duration_seconds, _last_prep_peak_rss_mb
 
     meta_qeff_model = _build_meta_qeff_model(qeff_model)
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/export.py]: The data type should be resolved once in export_weight_free_onnx() and passed down:
```
  target_dtype = qeff_model.model.config.dtype
  meta_qeff_model = _build_meta_qeff_model(qeff_model, target_dtype)
```

File: QEfficient/base/modeling_qeff.py (hunk 2 above)
Comment (amarquic) [reply QEfficient/base/modeling_qeff.py]: Resolved

File: QEfficient/utils/export_utils.py
Comment (amarquic) [reply QEfficient/utils/export_utils.py:1]: resolved

File: QEfficient/base/modeling_qeff.py (hunk 3 above)
Comment (amarquic) [reply QEfficient/base/modeling_qeff.py]: resolved

File: QEfficient/base/modeling_qeff.py (hunk 4 above)
Comment (amarquic) [reply QEfficient/base/modeling_qeff.py:499]: resolved.

File: QEfficient/base/modeling_qeff.py (hunk 5 above)
Comment (amarquic) [reply QEfficient/base/modeling_qeff.py]: resolved.

File: QEfficient/base/modeling_qeff.py (hunk 6 above)
Comment (amarquic) [reply QEfficient/base/modeling_qeff.py]: resolved

File: QEfficient/exporter/onnx_exporter.py (hunk 7 above)
Comment (amarquic) [reply QEfficient/exporter/onnx_exporter.py]: resolved

File: QEfficient/exporter/onnx_exporter.py (hunk 8 above)
Comment (amarquic) [reply QEfficient/exporter/onnx_exporter.py]: resolved

File: QEfficient/exporter/onnx_exporter.py (hunk 10 above)
Comment (amarquic) [reply QEfficient/exporter/onnx_exporter.py]: resolved

File: QEfficient/utils/torch_patches.py (hunk 11 above)
Comment (amarquic) [reply QEfficient/utils/torch_patches.py:427]: resolved

File: QEfficient/utils/cache.py (hunk 12 above)
Comment (amarquic) [reply QEfficient/utils/cache.py]: Resolved

File: QEfficient/exporter/weight_free/export.py (hunk 13 above)
Comment (amarquic) [reply QEfficient/exporter/weight_free/export.py]: resolved

File: QEfficient/exporter/weight_free/export.py (hunk 15 above)
Comment (amarquic) [reply QEfficient/exporter/weight_free/export.py]: resolved

File: QEfficient/exporter/weight_free/export.py (hunk 16 above)
Comment (amarquic) [reply QEfficient/exporter/weight_free/export.py]: done

File: QEfficient/exporter/weight_free/export.py (hunk 17 above)
Comment (amarquic) [reply QEfficient/exporter/weight_free/export.py]: Done

**Comment by quic-amitraj** [comment]
@amarquic Please apply ruff with the same version that is getting used in our repo. Except that all LGM.

**Review by quic-amitraj — APPROVED** [review]
